### ISSUE TITLE
Decide on metrics to compare performance of NDEs

### ISSUE THREAD

**Issue by EiffL** [astrodeepnet/sbi_experiments#17]

**Comment by Justinezgh** [comment]
In the paper [Benchmarking Simulation-Based Inference](https://arxiv.org/abs/2101.04653) they propose different metrics depending on the 'true information' and the algorithm used for the approximation. 

If we only have the 'true' observations and we assume that we has access to the ground truth parameters we can use :
- log likelihood (2)
- posterior predictive checks (1)
but it doesn't seem to be the best 

If we can sample from the ground truth posterior : 
- Maximum mean discrepancy (3)
- Classifier 2-Sample Tests (3)

If we have access to the gradients of the ground truth posterior : 
- Kernelized Stein Discrepancy (4)

If we have access to the ground truth posterior : 
- f-Divergences (5)

![image](https://user-images.githubusercontent.com/72011736/148532103-2c02ff86-afd7-4686-a1d2-660810bf52a9.png)
with q tilde the unnormalized posterior approximation and q the normalized one.

**Comment by Justinezgh** [comment]
If we only want to test different architectures maybe the f-divergences like the kl divergence is the best since we have access to the true distribution and the approximate one

**Comment by EiffL** [comment]
Yeah so I think in our toy experiments with analytic distributions like 2 moons, we could use all of them, and would be interesting to see how much an improvement of these metrics translates into a qualitative improvement of the posterior.

In practical cosmology applications, we'll see, it depends on the particular scenario we are considering. We could still be able to obtain samples and gradients of the ground truth, but expensive to obtain (we need to run an HMC).

From practical considerations, let's start with one, whichever one is simplest to implement. I guess all the codes can be found in Jan-Matthis' library, but it's all in pytorch ^^'

**Comment by Justinezgh** [comment]
Just to be sure, what we want to do first is to use all these metrics on different architectures to evaluate if the NFs learn the two moons distribution correctly ?

**Comment by EiffL** [comment]
yep, at stage 0, but we don't have to do all the metrics at once, we can start with one and try to make a plot.

The plot we want is number of sims on the x axis, quality metric on the y axis. And on this plot, you can have several lines, one per model/architecture/training loss used.

We'll ultimately want to have several of these plots, one per distribution we want to learn. We can start with 2 moons, but then we'll want to use some cosmology example.

**Comment by Justinezgh** [comment]
I tried to implement the dkl but I have negative values and a weird behavior : 
![Unknown](https://user-images.githubusercontent.com/72011736/148691120-ac384ebd-30f9-47b1-964b-e448f71b3bc2.png)

Here is my dkl if you see a mistake I'll be happy to know what it is ^^'

``` python 
model_NF = hk.without_apply_rng(hk.transform(lambda x : Flow()().log_prob(x)))
model_sample = hk.without_apply_rng(hk.transform(lambda nb_sample: Flow()().sample(nb_sample, seed=next(rng_seq))))

def kl_divergence_fn(params, distribution_b, precision):      
  z = model_sample.apply(params, precision)        
  return jnp.mean(model_NF.apply(params,z) - distribution_b.log_prob(z))
```

**Comment by EiffL** [comment]
ouuuuuh that's really cool :-D 

On the x axis, you have the number of samples you use to evaluate the metric, right?  So what we are seeing is that it needs 100000 samples to converge to a given value?

**Comment by Justinezgh** [comment]
hmm maybe I'm confused but I think that's not cool ^^
Yes the x axis is the nb of simulations and the y axis is the kl divergence (which is supposed to be close to zero if the approximation is good right ?)
Plus, this is the contour plot for the nll loss : 

![nll-2](https://user-images.githubusercontent.com/72011736/148693233-b48f0f94-6bc4-4da5-9324-3ad2220bf629.png)
 
and the one for the score/2000 + nll loss :
![Unknown-3](https://user-images.githubusercontent.com/72011736/148693250-7800c84c-3fbc-4a15-9988-032c7bc7b525.png)

so we see that the nll loss gets better results faster but I think that this is not what we see with the dkl plot

**Comment by Justinezgh** [comment]
This is the notebook : [link](https://colab.research.google.com/drive/14_l5fiiUrzNq_IPDZp7CxS6ahdVcmUn1?usp=sharing)

I also have the one for the mmd metric : 
![Unknown](https://user-images.githubusercontent.com/72011736/148693535-fd20c578-fad2-4d55-a338-a8ac1b54afbf.png)

**Comment by Justinezgh** [comment]
NF log_prob  >>> two moons log_prob, and I think the pb come from the 'tfd.Independent' in the output of the NF

**Comment by EiffL** [comment]
hummmm yeah..... but that's what tfd.independent is supposed to do no? it sums over the dimensions that are not batched

**Comment by Justinezgh** [comment]
hmm yes right 
I don't know where the pb is

**Comment by EiffL** [comment]
but so, what kind of values do you have for the flow log prob values? I'm surprised if they are very big, because the NLL loss if I remember correctly gives you like -1.5 at most, that would be the average log prob for samples drawn from the two moons

**Comment by Justinezgh** [comment]
yes it was because I did that 

```python 
model_NF.apply(params, batch)
```

instead of 
```python
jax.vmap((lambda x : model_NF.apply(params, x.reshape([1,2])).squeeze()))(batch)
``` 
still a bit lost with Jax :)

**Comment by EiffL** [comment]
![image](https://user-images.githubusercontent.com/861591/149033676-94920a25-6477-4d68-a021-1a6e79c1aa03.png)
Lol, just got this ^^^^^

**Comment by EiffL** [comment]
yeah... so... something wrong in the way our code for the flow handles the batch dimension :-| ya..., not great. but at least if we vmap it works ^^

**Comment by EiffL** [comment]
it might be a problem in how the coupling layer is implemented, idk

**Comment by Justinezgh** [comment]
batch_size = [50,70,90,110]

- Nll
![Unknown-5](https://user-images.githubusercontent.com/72011736/149154876-7fb6ad26-c767-4d9f-a190-943240c48c09.png)

- Nll+score 
![Unknown-6](https://user-images.githubusercontent.com/72011736/149154929-5a50dcb0-a647-4a3f-a18a-a42db1bf3111.png)


![Unknown](https://user-images.githubusercontent.com/72011736/149154663-6c96bcf1-2090-497c-a525-812e1dd4cc06.png)

![Unknown-2](https://user-images.githubusercontent.com/72011736/149154674-0e4e73cd-0b58-474b-a4e1-d9685d0fee63.png)

![Unknown-3](https://user-images.githubusercontent.com/72011736/149154699-3ffd3ab3-8370-41d3-b87f-7366c0751490.png)

![Unknown-4](https://user-images.githubusercontent.com/72011736/149154713-14f0fd37-6282-4444-a8ad-dfc11854687f.png)

**Comment by EiffL** [comment]
Did you get better results without the outlier at N=70?

**Comment by Justinezgh** [comment]
Hmm not really, this is the best I have for now : 

Loss = nll
![nll](https://user-images.githubusercontent.com/72011736/149515487-3cc6572a-7431-4b12-9d03-9a308175daa1.png)

Loss = nll + score/1000
![nllscore](https://user-images.githubusercontent.com/72011736/149515498-1112a0ea-2842-476c-88de-75e92dba88f1.png)

Metrics :
![c2st](https://user-images.githubusercontent.com/72011736/149515447-da49b509-b00a-4e8d-86bb-e2e277a9b1c4.png)
![dkl](https://user-images.githubusercontent.com/72011736/149515565-1a3ad957-d765-4082-a68a-ea8e668168b5.png)
![nlp](https://user-images.githubusercontent.com/72011736/149515576-083d0067-52ec-4ee6-854e-80e35f3ccc4d.png)![mmd](https://user-images.githubusercontent.com/72011736/149515594-6a076387-b46f-4102-9c9f-c41733f2a4d6.png)


I changed the variance of the normal distribution for the latent variable to avoid nan in the loss (was set to 1): 
``` python 
      nvp = tfd.TransformedDistribution(
            tfd.Independent(tfd.TruncatedNormal(0.5*jnp.ones(d), 
                                                0.05*jnp.ones(d), 
                                                0.01,0.99),
                            reinterpreted_batch_ndims=1),
            bijector=chain)
```

and I added randomness during training : 
``` python 
    batch, score_batch = get_batch(jax.random.PRNGKey(0))
    for step in tqdm(range(5000)):
      inds = np.random.randint(0, batch_size, int(batch_size / 2))
      l, params, opt_state = update(params, opt_state, batch[inds], score_batch[inds])
```
